<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 1: Python Coding Exercises

_Authors: Joseph Nelson (DC) _

---

The following code challenges are drawn from common exercises used in technical interviews.

Please note that there may be several ways to approach each challenge. If you get stuck, try mapping out your approach in pseudocode first. Finally, while solutions to problems like these may be found online, remember that if you copy/paste code that you can't explain, you'll be missing out on the point of the project. The only way to truly learn a new skill is through practice, trial, and error - we can only help you improve by understanding where you are having trouble.

### Challenge 1: Largest Palindrome
A palindromic number reads the same both ways. For example, 1234321 is a palindrome. The largest palindrome made from the product of two two-digit numbers is 9009 = 91 × 99. Find the largest palindrome made from the product of two three-digit numbers. Afterward, write a brief explanation walking through your code's logic in markdown.

In [1]:
print("My First Challenge code")

My First Challenge code


In [5]:
num = 0
for n1 in range(999, 100, -1):
    for n2 in range(n1, 100, -1):
        x = n1 * n2
        if x > num:
            s = str(n1 * n2)
            if s == s[::-1]:
                num = n1 * n2
print(num)

906609


The largest palindrome made from the product of two 3-digit numbers is 913 * 993 = 906609


### Challenge 2: Summation of Primes
The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17. Find the sum of all the primes below 2,000. Afterward, write a brief explanation walking through your code's logic in markdown.

In [7]:
print( "Challenger problem 2:")
num =2000
sum = 0

for i in range(2, num + 1):

    j = 2
    
    for j in range(2, i):
        if (int(i % j) == 0):
            j = i
            break;

    #If the number is prime then add it.
    if j is not i:
        sum += i

print("\nSum of all prime numbers upto", num, ":", sum)

Challenger problem 2:

Sum of all prime numbers upto 10 : 15


### Challenge 3: Multiples of 3 and 5
If we list all of the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6, and 9. The sum of these multiples is 23. Find the sum of all the multiples of 3 and 5 below 1,000. Afterward, write a brief explanation walking through your code's logic in markdown.

In [22]:
nums =[]
for i in range (3,100) :
    #print (i)
    if i%3==0 and  i%5==0 :
        if i not in nums :
            nums.append(i)
print(len(nums))        
sum = 0
n = len(nums)

print(nums)
for i in range (0,n) :
    print(sum)
    sum =sum + nums[i]
print(sum )   

6
[15, 30, 45, 60, 75, 90]
0
15
45
90
150
225
315


### Challenge 4: String Compressor
Implement a method to perform basic string compression using the counts of repeated characters. (This is called run-length encoding.) For example, the string "aabcccccaaa" would become a2b1c5a3. If the “compressed” string would not become smaller than the original string, your method should return the original string. You can assume the string has only uppercase and lowercase letters (a–z). Specify whether your solution is case sensitive or case insensitive and what you would need to change to make it the other. Afterward, write a brief explanation walking through your code's logic in markdown.

In [5]:
def string_compress(string):
    out_str= ""
    count= 0
    currentchar = ""
    
    for char in string :
        if char== currentchar :
            count+=1
        else:
            if count >0 :
                out_str += currentchar +str(count)
            currentchar = char
            count =1
    out_str += currentchar +str(count)     
    
    return(out_str)

print (string_compress('Anjali'))
print (string_compress('aabcccccaaa'))
        
    

 
    


A1n1j1a1l1i1
a2b1c5a3


### *BONUS* Challenge: FizzBuzz
Write a program that prints all of the numbers from 1 to 100. For multiples of 3, instead of the number, print "Fizz;" for multiples of 5, print "Buzz." For numbers that are multiples of both 3 and 5, print "FizzBuzz." Afterward, write a brief explanation walking through your code's logic in markdown.

In [16]:
num = 100
for i in range ( 1,101) :
    if (i%3==0 and  i%5==0):
        print('number is:{} , and its FizzBuzz'.format(i))
    elif i%3== 0 :
        print('number is:{} , and its Fizz'.format(i))
    elif i%5==0 : 
        print('number is:{} , and its Buzz'.format(i))

number is:3 , and its Fizz
number is:5 , and its Buzz
number is:6 , and its Fizz
number is:9 , and its Fizz
number is:10 , and its Buzz
number is:12 , and its Fizz
number is:15 , and its FizzBuzz
number is:18 , and its Fizz
number is:20 , and its Buzz
number is:21 , and its Fizz
number is:24 , and its Fizz
number is:25 , and its Buzz
number is:27 , and its Fizz
number is:30 , and its FizzBuzz
number is:33 , and its Fizz
number is:35 , and its Buzz
number is:36 , and its Fizz
number is:39 , and its Fizz
number is:40 , and its Buzz
number is:42 , and its Fizz
number is:45 , and its FizzBuzz
number is:48 , and its Fizz
number is:50 , and its Buzz
number is:51 , and its Fizz
number is:54 , and its Fizz
number is:55 , and its Buzz
number is:57 , and its Fizz
number is:60 , and its FizzBuzz
number is:63 , and its Fizz
number is:65 , and its Buzz
number is:66 , and its Fizz
number is:69 , and its Fizz
number is:70 , and its Buzz
number is:72 , and its Fizz
number is:75 , and its FizzBuzz
numb